### PR TITLE
init docker workflow

### LINKED ISSUE #5: Add Dockerfile and build pipeline for CI
- the service should be shipped as docker container,  for build we should use multistage image with `scratch` as final step
- On PR  we should verify that build is successful 
- on Merge image should be builded and published to GitHub packages(docker registry)

### PR DESCRIPTION
**Pull request by shan-96**
closes #5 

- add workflow and required settings to build and deploy docker images

### CODE REVIEW

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@v4
+        with:
+          images: ghcr.io/${{ github.repository }}
+
+      - name: Build and push Docker image
```
Comment (shan-96) [inline .github/workflows/docker.yml:38]: @ksysoev I need to credentials setup to push to github registry. Could you help?
image

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+WORKDIR /app
+
+COPY go.mod go.sum ./
+
+RUN go mod download
+
+COPY . .
+
+RUN go build -o /app/cmd/bff/cmd ./cmd/bff
```
Comment (shan-96) [inline Dockerfile:11]: what modules would you like to add, and keep as entrypoint?

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -0,0 +1,71 @@
+
+name: Docker CI/CD
+
+on:
+  push:
+    branches:
+      - main
+    tags:
+      - '*'
+  pull_request:
```
Comment (shan-96) [inline .github/workflows/docker.yml:10]: should we run on each commit of pull request?

Comment (ksysoev) [reply .github/workflows/docker.yml:10]: @shan-96 
- I think for pull request we should build, but not push, to be sure that we're not merging failing build
- on every comment to master we should build and push with tag latest
- on release/tag we should build and push with tag name, that will contain version number.

make sense?

File: .github/workflows/docker.yml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+          echo "IMAGE_TAG=${{ github.ref_name }}" >> .env
+          echo "CONTAINER_NAME=deriv-api-bff-${{ github.ref_name }}" >> .env
+          echo "HOST_PORT=8080" >> .env
+          echo "GLOBAL_DEBUG_ON=true" >> .env
+
+      - name: Start Docker Compose
+        run: ./scripts/start_container.sh
+
+      - name: Stop Docker Compose
+        run: ./scripts/stop_container.sh
```
Comment (ksysoev) [inline .github/workflows/docker.yml:71]: @shan-96 I think it'll never fail, I think these 2 steps doesn't provide a value

File: .github/workflows/docker.yml (hunk 1 above)
Comment (ksysoev) [reply .github/workflows/docker.yml:38]: @shan-96 gave write permissions for workflow, hope it should work.

File: Dockerfile (hunk 2 above)
Comment (ksysoev) [reply Dockerfile]: let's keep out put binary as entry point and server command as entry point

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+RUN go mod verify
+
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o /main ./cmd/bff
+
+FROM scratch
+
+COPY --from=builder /usr/share/zoneinfo /usr/share/zoneinfo
+COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
+COPY --from=builder /etc/passwd /etc/passwd
+COPY --from=builder /etc/group /etc/group
```
Comment (ksysoev) [inline Dockerfile:26]: @shan-96
1) here just lack of my knowledge do we need zoneinfo, passwd and group? what will be broken if we don't copy that? 
2) shall we copy default config with no defined calls and ship it in image by default? to be able to start server in pass-trough mode by default. now it'll be crashing if we don't mount config to the container

Comment (shan-96) [reply Dockerfile:26]: 1. yes we do not need to create default user. I can remove it. Probably even the certs
2. sure, but lets handle that in docker-compose

File: .github/workflows/docker.yml (hunk 3 above)
Comment (shan-96) [reply .github/workflows/docker.yml:10]: lets follow this style 
image

File: .github/workflows/docker.yml (hunk 4 above)
Comment (shan-96) [reply .github/workflows/docker.yml:73]: yes right now just focusing on starting and stopping the server

File: .github/workflows/docker.yml (hunk 1 above)
Comment (shan-96) [reply .github/workflows/docker.yml:38]:
```
Error: buildx failed with: ERROR: failed to solve: failed to push ghcr.io/ksysoev/deriv-api-bff:pr-10: unexpected status from POST request to https://ghcr.io/v2/ksysoev/deriv-api-bff/blobs/uploads/: 403 Forbidden
```

this is the latest error. earlier it was 401

Comment (shan-96) [reply .github/workflows/docker.yml:42]: @ksysoev I think I found the problem, I am trying to login using
```
name: Log in to GitHub Container Registry
uses: docker/login-action@v2
with:
    registry: ghcr.io
    username: ksysoev
    password: ${{ secrets.GITHUB_TOKEN }}
```

I think GITHUB_TOKEN is not set as a CI CD secret

Comment (ksysoev) [reply .github/workflows/docker.yml:42]: @shan-96 should not be the case, `secrets.GITHUB_TOKEN` should be there by default, if  workflow is configured 
image

Comment (ksysoev) [reply .github/workflows/docker.yml:42]: @shan-96 based on documentation, you config file looks to correct, if it'll not work. lets just add build for now I'll try to debug push to registry separately.  Maybe I need to configure something else as well... I never used GitHub container registry before.

File: .github/workflows/docker.yml (hunk 3 above)
Comment (ksysoev) [reply .github/workflows/docker.yml:10]: looks good

File: .github/workflows/docker.yml (hunk 1 above)
Comment (shan-96) [reply .github/workflows/docker.yml:42]: @ksysoev another doubt, is this workflow permission same for all jobs?

**Pull request merged**